### PR TITLE
implement Countdown component

### LINKED ISSUE #4: Countdown
screen shot 2017-09-20 at 7 20 50 pm

### PR DESCRIPTION
**Pull request by R1ZZU**
Closes #4

### CODE REVIEW

File: src/components/Countdown/Countdown.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import { h, Component } from 'preact';
+
+const COMPONENT_SIZE_TO_FONT_SIZE_RATIO = 4;
+
+class Countdown extends Component {
+    shouldComponentUpdate = () => false
+    render = () => <canvas ref={canvas => this.canvas = canvas} />
```
Comment (DaQuirm) [inline src/components/Countdown/Countdown.js]: YO DAWG

File: src/components/Countdown/Countdown.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+        this.ctx.textBaseline = 'middle'
+        this.ctx.lineCap = 'round';
+    }
+
+    _clear() {
+        this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
+    }
+
+    _render({ size, fontSize, arcColor, remainigTimeArcColor, timeAmount, timeRemaining }) {
```
Comment (DaQuirm) [inline src/components/Countdown/Countdown.js]: Here and below (see stories too): typo `remainig`

**Comment by DaQuirm** [comment]
Do we need to do some kind of teardown (e.g. for `ctx`) on unmount?

File: src/components/Countdown/Countdown.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        this._clear();
+        this._renderText({ timeRemaining });
+
+        this.ctx.strokeStyle = arcColor;
+        this._renderArc({ progress: 1 });
+
+        this.ctx.strokeStyle = remainigTimeArcColor;
+        this._renderArc({
```
Comment (kirilknysh) [inline src/components/Countdown/Countdown.js]: Am I correct, that the progress is updated on every `componentWillReceiveProps`? Which means every 1 second and that would be "jumping" update: I mean update from 94 seconds to 93 seconds is a "jump", not animation of 1 second duration.

**Comment by R1ZZU** [comment]
> Do we need to do some kind of teardown (e.g. for ctx) on unmount?
@DaQuirm at first glance this question seems reasonable and I'd say "Yes, sure, my bad". However as we don't have any DOM event listeners bound to canvas, there is no need to do any kind of cleanup. Canvas will be removed from DOM and rendering context will be destroyed as well.

**Comment by DaQuirm** [comment]
> at first glance this question seems reasonable
I really hope it _is_ , in any case, a reasonable question 😺 
> there is no need to do any kind of cleanup
Excellent! Just what I wanted to know 👍

File: src/components/Countdown/Countdown.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+        const newProgress = this.progress;
+
+        if (nextProps.timeRemaining === nextProps.timeAmount) {
+            this._render();
+        } else {
+            this._animate({
+                from: prevProgress,
+                to: newProgress,
+                duration: 1000,
```
Comment (kirilknysh) [inline src/components/Countdown/Countdown.js]: `(newProgress - prevProgress) * 1000` ?

Comment (R1ZZU) [reply src/components/Countdown/Countdown.js]: Good point, I had this implementation initially, _but_... What if for some reason (not sure if that is possible at all) user will miss one message about clock tick (e.g. it was 40 seconds remaining, then – 38). In this case we'll have 2 secs animation which will be interrupted after next tick message. This looks not really good, I'd leave 1 sec animation for all cases

Comment (R1ZZU) [reply src/components/Countdown/Countdown.js]: On the other hand if we'll need to notify about timer progress more frequently than once per second we'll have troubles...

Comment (R1ZZU) [reply src/components/Countdown/Countdown.js]: Ideas?

Comment (DaQuirm) [reply src/components/Countdown/Countdown.js]: I had one, but maybe too epic for now: We update the countdown based on an interval set on the client and only use the countdown messages for correction (like atomic clock sync ⌚️ ).

Comment (kirilknysh) [reply src/components/Countdown/Countdown.js]: Let's assume `NORMAL_DURATION` is `1000`.
```
if (newProgress - prevProgress > NORMAL_DURATION) {
    return NORMAL_DURATION;
} else {
    return newProgress - prevProgress;
}
```

Comment (R1ZZU) [reply src/components/Countdown/Countdown.js]: Or.. Or... Or.....
We can store previous update time and use it as baseline for animation duration calculation. WDYT?

Comment (kirilknysh) [reply src/components/Countdown/Countdown.js]: The component looks over complicated already... let's keep it 1000ms for now...

Comment (R1ZZU) [reply src/components/Countdown/Countdown.js]: >  looks over complicated already...

That's not good 😞 I'd see what can I do in order to simplify it

**Pull request merged**